### ISSUE TITLE
#32695 breaks aframe in DT

### ISSUE THREAD

**Issue by sandersn** [microsoft/TypeScript#33582]
https://github.com/microsoft/TypeScript/pull/32695

1. clone DT
2. cd ~/dt/types/aframe
3. code .
4. Bogus error on setupFadeAnimation on line 51 of tests/aframe-io-tests.ts; the property was not previously marked as excess, but is now marked as excess.

It looks like the intent of registerComponent is to allow excess properties for its second parameter, but it does so by intersecting a type parameter `T extends object` with `Partial<Component>`. This might be what changed.

```ts
AFRAME.registerComponent('set-image', {
	schema: {
		on: { type: 'string' },
		target: { type: 'selector' },
		src: { type: 'string' },
		dur: { type: 'number', default: 300 }
	},

	init: function() {
		var data = this.data;
		var el = this.el;

		this.setupFadeAnimation();

		el.addEventListener(data.on, function() {
			// Fade out image.
			data.target.emit('set-image-fade');
			// Wait for fade to complete.
			setTimeout(function() {
				// Set image.
				data.target.setAttribute('material', 'src', data.src);
			}, data.dur);
		});
	},

	/**
	 * Setup fade-in + fade-out.
	 */
	setupFadeAnimation: function() {
		var data = this.data;
```

The same error is also on lines 166, 910, 1025, and on line 86 of aframe-tests.ts

**Comment by sandersn** [comment]
vfile is also broken; here, `isCustomData` asserts that `VFile.data` is of type CustomData, not `unknown`. But it now fails to do so.

```ts
interface CustomData {
    message: string;
}
function isCustomData(data: any): data is CustomData {
    return data != null && typeof data.message === 'string';
}
function doSomethingWithCustomData(data: CustomData) {
}
const customData: CustomData = {
    message: 'tango'
};
const fileWithData = vfile({
    data: customData
});

if (isCustomData(fileWithData.data)) {
    doSomethingWithCustomData(fileWithData.data);
}
doSomethingWithCustomData(fileWithData.data); // $ExpectError
```

**Comment by sandersn** [comment]
node/v10 is broken, on fairly simple code:

```ts
        const v4 = fs.realpathSync.native('/path/to/folder', { encoding: s });
        typeof v4 === "string" ? s = v4 : b = v4;
```

where `v4: String | buffer` and `s: string` and `b: Buffer`.

weixin-app is broken, probably for the same reason as aframe; it's a Vue-like type with an intersection that includes ThisType.

**Comment by ahejlsberg** [comment]
A couple of things going on.

In node/v10 and vfile there are calls to never-returning functions that now cause unreachable code errors. It's a simple fix to the tests.

The aframe issue is more involved but basically comes down to calls of the form `this.foo(...)` being subject to control flow analysis in order to determine end-point reachability in functions. This causes use to more eagerly resolve a `ThisType<...>` which in turn causes type inferences to be ignored. The fix here is to ensure only `this.foo(...)` calls in methods of classes or functions with explicit `this` parameters having explicit type annotations are analyzed (similar to what we do for identifiers).

I'll put up some PRs.
